lib/ansible/modules/crypto/openssh_keypair.py

```diff
@@ -0,0 +1,351 @@
+    type: string
+    sample: 4096 SHA256:r4YCZxihVjedH2OlfjVGI6Y5xAYtdCwk8VxKyzVyYfM example@example.com (RSA)
+'''
+
+import os
+import stat
+import errno
+import subprocess
+import tempfile
+import codecs
+
+from subprocess import PIPE, Popen
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from pipes import quote
+
+
+class PrivateKeyError(Exception):
+    pass
+
+
+class PrivateKey(object):
+
+    def __init__(self, module):
+        self.path = module.params['path']
+        self.state = module.params['state']
+        self.force = module.params['force']
+        self.size = module.params['size']
+        self.type = module.params['type']
+        self.passphrase = module.params['passphrase']
+        self.comment = module.params['comment']
+        self.changed = False
+        self.check_mode = module.check_mode
+        self.privatekey = None
+        self.fingerprint = {}
+
+        if self.type in ('rsa', 'rsa1'):
+            self.size = 4096 if self.size is None else self.size
+            if self.size < 1024:
+                module.fail_json(msg='''For RSA keys, the minimum size is 1024 bits and the default is 4096 bits.
+                                     Attempting to use bit lengths under 1024 will cause the module to fail.''')
+
+        if self.type == 'dsa':
+            self.size = 1024 if self.size is None else self.size
+            if self.size != 1024:
+                module.fail_json(msg='''DSA keys must be exactly 1024 bits as specified by FIPS 186-2.''')
+
+        if self.type == 'ecdsa':
+            self.size = 256 if self.size is None else self.size
+            if self.size not in (256, 384, 521):
+                module.fail_json(msg='''For ECDSA keys, size determines the key length by selecting from
+                                     one of three elliptic curve sizes: 256, 384 or 521 bits.
+                                     Attempting to use bit lengths other than these three values for
+                                     ECDSA keys will cause this module to fail.''')
+        if self.type == 'ed25519':
+            self.size = 256
+
+    def generate(self, module):
+        # generate a keypair
+        if not self.check(module, perms_required=False) or self.force:
+            args = [
+                'ssh-keygen',
+                '-q',
+                '-b', str(self.size),
+                '-t', self.type,
+                '-f', self.path,
+            ]
+            try:
+                tmpfile = None
+                try:
+                    if self.passphrase:
+                        with tempfile.NamedTemporaryFile(delete=False) as tf:
+                            tmpfile = tf.name
+                        os.chmod(tmpfile, stat.S_IRWXU)
+                        with codecs.open(tf.name, "w", encoding='utf-8') as tf:
+                            tf.write("#!/bin/sh\necho %s" % quote(self.passphrase))
```

Comment:
In the "Generate privatekey6 - standard - with non-ASCII passphrase" test, this fails with `'ascii' codec can't decode byte 0xc3 in position 16: ordinal not in range(128)`. Stacktrace:
```
File "/tmp/ansible_openssh_keypair_payload_kCnMbD/__main__.py", line 193, in generate
  tf.write("#!/bin/sh\necho %s" % quote(self.passphrase))
File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/codecs.py", line 706, in write
  return self.writer.write(data)
File "/usr/local/Cellar/python/2.7.12/Frameworks/Python.framework/Versions/2.7/lib/python2.7/codecs.py", line 369, in write
  data, consumed = self.encode(object, self.errors)
```
It seems to fail in all Python 2 tests; in Python 3 it seems to work.

Reply:
it works either in py 3 or py 2 i haven't found a way to get it to work with both versions yet